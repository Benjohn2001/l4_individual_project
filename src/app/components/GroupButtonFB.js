import { Component, useState } from "react";
import React from "react";
import PropTypes from 'prop-types';
import { TouchableOpacity, Text, View, Image, FlatList, ActivityIndicator } from "react-native";
import { Feather } from '@expo/vector-icons';
import PressableIcon from "./PressableIcon";
import { getStorage, getDownloadURL, uploadBytesResumable, uploadBytes, uploadString } from "firebase/storage";
import {ref as stref} from "firebase/storage";
import { query, get, ref, getDatabase } from 'firebase/database';
import { auth } from "../../firebase";

function GroupButtonFB (props) {

    const { groupName, onPress, membRef} = props;
    const [pics, setPics] = useState([]);
    const [urls, setUrls] = useState([])
    const [membersKeys, setMembersKeys] = useState([])
    const [fetched, setFetched]=useState(false)

    const uid=auth.currentUser.uid

    React.useEffect(() =>{
        fetchData()
    },[] )

    const fetchData = async () =>{
        setFetched(false)
        setMembersKeys([])
        setUrls([])
        setPics([])
        const groupMembersRef = query(ref(getDatabase(), "/groupMembers/"+membRef))
        const data = await get(groupMembersRef)
        for (const c in data.val()){
            if(data.val()[c]["member"]!==uid){
                const refV = query(ref(getDatabase(), "/users/"+data.val()[c]["member"]))
                const dataV= await get(refV)
                getDownloadURL(stref(getStorage(),dataV.val()["profilePic"])).then((url)=>{
                    setPics(a => {return [...a , url]})
                })
            }
            
        }
        setFetched(true)
    }
    
    return (
        <View className="items-center w-full">
        <View className="py-3 w-11/12 bg-primaryDarker h-36 rounded-2xl">
            <TouchableOpacity onPress={onPress} >
            <View className="flex-row w-11/12 pb-3">
                <Text  className="text-black text-2xl text-center ml-5 mr-3">{groupName}</Text>
                <View  className="ml-auto mr-5">
                    <Feather
                        name="arrow-right"
                        size={32}
                        color="black"
                    /> 
                </View>  
            </View>
            </TouchableOpacity>
            <View className="items-center">
            {fetched ? 
                    <FlatList
                        showsHorizontalScrollIndicator={false}
                        horizontal={true}
                        data={pics}
                        extraData={pics}
                        renderItem={ ({item})=>(
                            <View className="mx-1">
                                <Image 
                                    className="w-16 h-16 rounded-full" 
                                    source = {{uri: item}}
                                />
                            </View>
                        )}
                    />
                :
                    <ActivityIndicator size="large" color="#6B4EFF"  />
                }
            </View>

        </View>
        </View>
        
    );
}

GroupButtonFB.propTypes = {
    groupName: PropTypes.string.isRequired,
    onPress: PropTypes.func.isRequired,
    membRef: PropTypes.string.isRequired
};

export default GroupButtonFB;