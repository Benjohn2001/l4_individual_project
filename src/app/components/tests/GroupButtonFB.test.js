import { render, screen, fireEvent } from "@testing-library/react-native";
import renderer from "react-test-renderer";
import GroupButtonFB from "../GroupButtonFB";

jest.mock("firebase/database", () => ({
  getDatabase: jest.fn().mockReturnThis(),
}));

jest.mock("firebase/storage", () => ({
  getStorage: jest.fn().mockReturnThis(),
}));

jest.mock("firebase/app", () => ({
  auth: jest.fn().mockReturnThis(),
  initializeApp: jest.fn().mockReturnThis(),
}));

jest.mock("firebase/auth", () => ({
  getAuth: jest.fn().mockReturnThis(),
}));

test("Check GroupButtonFB renders correctly and onpress works", () => {
  const onPressMock = jest.fn();
  render(
    <GroupButtonFB
      groupName="press"
      onPress={() => {
        onPressMock();
      }}
      membRef="aaaaa"
    />
  );
  fireEvent.press(screen.getByText("press"));
  expect(onPressMock);
  const domTree = renderer
    .create(
      <GroupButtonFB
        groupName="press"
        onPress={() => {
          onPressMock();
        }}
        membRef="aaaaa"
      />
    )
    .toJSON();
  expect(domTree).toMatchSnapshot();
});
