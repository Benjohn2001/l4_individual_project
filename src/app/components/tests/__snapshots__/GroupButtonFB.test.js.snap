// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check GroupButtonFB renders correctly and onpress works 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
      },
      Object {
        "width": "100%",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "paddingBottom": 12,
          "paddingTop": 12,
        },
        Object {
          "width": "91.666667%",
        },
        Object {
          "backgroundColor": "#9990FF",
        },
        Object {
          "height": 144,
        },
        Object {
          "borderBottomLeftRadius": 16,
          "borderBottomRightRadius": 16,
          "borderTopLeftRadius": 16,
          "borderTopRightRadius": 16,
        },
      ]
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
            },
            Object {
              "width": "91.666667%",
            },
            Object {
              "paddingBottom": 12,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#000",
              },
              Object {
                "fontSize": 24,
                "lineHeight": 32,
              },
              Object {
                "textAlign": "center",
              },
              Object {
                "marginLeft": 20,
              },
              Object {
                "marginRight": 12,
              },
            ]
          }
        >
          press
        </Text>
        <View
          style={
            Array [
              Object {
                "marginLeft": "auto",
              },
              Object {
                "marginRight": 20,
              },
            ]
          }
        >
          <Text />
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
          },
        ]
      }
    >
      <ActivityIndicator
        color="#6B4EFF"
        size="large"
      />
    </View>
  </View>
</View>
`;
